sqrl-engines/sqrl-engine-flink/src/main/java/com/datasqrl/engine/stream/flink/plan/SqrlToFlinkExecutablePlan.java

```diff
@@ -197,51 +193,57 @@ private RelNode applyDowncasting(RelNode relNode, Optional<ExecutionEngine> engi
+    //1. Check to see if format or engine supports converting the data type to a type flink can handle natively
+    TableConfig tableConfig = getTableConfig(writeSink);
+    if (!(field.getType() instanceof RawRelDataType) ||
+        formatSupportsType(tableConfig.getFormat().getName(), (RawRelDataType)field.getType()) ||
+        engineSupportsType(getEngine(writeSink), ((RawRelDataType) field.getType()).getRawType().getDefaultConversion())) {
+      return relBuilder.field(field.getIndex());
+    }
 
-      FlinkConverter flinkConverter = new FlinkConverter((TypeFactory) framework.getQueryPlanner().getCatalogReader()
-          .getTypeFactory());
+    //Otherwise apply downcasting
+    Class<?> defaultConversion = ((RawRelDataType) field.getType()).getRawType()
+        .getDefaultConversion();
+    CatalogReader catalogReader = (CatalogReader) relBuilder.getRelOptSchema();
 
-      Optional<SqlFunction> convertedFunction = flinkConverter
-          .convertFunction(fncName, functionDef);
+    DowncastFunction downcastFunction = ServiceLoaderDiscovery.get(DowncastFunction.class,
+        e -> e.getConversionClass().getName(),
+        defaultConversion.getName());
+    if (downcastFunction == null) {
+      throw new RuntimeException("Needed downcast function but could not find one: " + defaultConversion.getName());
+    }
+    String fncName = downcastFunction.downcastFunctionName().toLowerCase();
+    FunctionDefinition functionDef;
+    try {
+      functionDef = (FunctionDefinition)downcastFunction.getDowncastClassName()
+          .getDeclaredConstructor().newInstance();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    SqrlFramework framework = catalogReader.getSchema().getSqrlFramework();
 
-      if (convertedFunction.isEmpty()) {
-        throw new RuntimeException("Could not convert downcast function");
-      }
+    FlinkConverter flinkConverter = new FlinkConverter((TypeFactory) framework.getQueryPlanner().getCatalogReader()
+        .getTypeFactory());
 
-      downcastClassNames.put(fncName,
-          downcastFunction.getDowncastClassName().getName());
+    Optional<SqlFunction> convertedFunction = flinkConverter
+        .convertFunction(fncName, functionDef);
 
-      hasChanged.set(true);
-      return relBuilder.getRexBuilder()
-          .makeCall(convertedFunction.get(), List.of(relBuilder.field(field.getIndex())));
+    if (convertedFunction.isEmpty()) {
+      throw new RuntimeException("Could not convert downcast function");
     }
 
-    return relBuilder.field(field.getIndex());
+    downcastClassNames.put(fncName,
+        downcastFunction.getDowncastClassName().getName());
+
+    hasChanged.set(true);
+    return relBuilder.getRexBuilder()
+        .makeCall(convertedFunction.get(), List.of(relBuilder.field(field.getIndex())));
+  }
+
+  private boolean engineSupportsType(Optional<ExecutionEngine> engine, Class<?> defaultConversion) {
```

Comment:
Just some food for thought: I see that you are tying the determination of downcasting to the engine. From the perspective of Flink, I wonder if we should make that determination for the connector.
In other words, does it matter whether we are writing to Postgres as an engine or to Postgres as an external connector?

Reply:
yeah, we can move to a connector + format when we transition away from the connectors that are streams. Right now it's a lot of code to migrate.